clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -842,8 +843,8 @@ public void seekToEnd(Collection<TopicPartition> partitions) {
 
         acquireAndEnsureOpen();
         try {
-            Collection<TopicPartition> parts = partitions.isEmpty() ? subscriptions.assignedPartitions() : partitions;
-            subscriptions.requestOffsetReset(parts, OffsetResetStrategy.LATEST);
+            cachedSubscriptionHasAllFetchPositions = false;
+            applicationEventHandler.add(new ResetOffsetEvent(partitions, OffsetResetStrategy.LATEST));
```

Comment:
`seekToBeginning` and `seekToEnd` don't wait the completion, but `seek` does. not sure why those similar methods have different behavior. The doc of `seek` declares the effect will be applied on next `poll`, and `ClassicKafkaConsumer#seek` does not wait also. IMHO, `AsyncKafkaConsumer#seek` should use `add` rather than `addAndGet`.

https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L792

@kirktrue @lianetm WDYT?

Reply:
Hey here, sorry for the late reply. I agree that these 3 methods should have similar behaviour, but I wonder if the case is that they should all use `addAndGet`? My concern is, when seeking (either of the 3 methods), we do expect that as soon as they complete, the position is updated so that other funcs will see it (ex. currentLag). We would be not respecting that anymore if we use add. seek could complete without having updated the offsets right?

From the seek contract, it states that "Overrides the fetch offsets that the consumer will use on the next {@link #poll(Duration) poll(timeout)}", we we need to:
1. override fetch offsets (this requires `addAndGet` I expect)
2. make sure they are available on the next poll (again addAndGet on seek, otherwise consumer.seek + consumer.poll may not find the offsets if the poll happens when the `seek` with `add` is still being processed in the background right?)

Thoughts?

Reply:
@lianetm thanks for the sharing. you are right, those seek methods need to wait the event gets completed to avoid race condition. +1 to use `addAndGet` on `seek`, `seekToBeginning`, and `seekToEnd`. I will close https://issues.apache.org/jira/browse/KAFKA-17586

Reply:
@TaiJuWu Could you please use `addAndGet` instead?

Reply:
Another option we've used elsewhere is to stash a reference to the `ResetOffsetsEvent` in the `AsyncKafkaConsumer` after it's enqueued so it can be checked later (i.e. in `poll()`).

I will admit I don't like the idea 🤔

Reply:
Hi @chia7712 @lianetm , thanks for your review and sharing.
I will update this PR according your suggestion.
Thanks a lot.

Reply:
I noticed another difference between `seek()`, `seekToBeginning()`, and `seekToEnd()`—`seek()` sets the position and validates it (if possible) while the other two forms _request_ the positions be reset. So even if the code in `seekToBeginning()` and `seekToEnd()` is switched to `addAndGet()`, it doesn't guarantee that the positions have been reset when the method exits, does it?

Reply:
> I noticed another difference between `seek()`, `seekToBeginning()`, and `seekToEnd()`—`seek()` sets the position and validates it (if possible) while the other two forms _request_ the positions be reset. So even if the code in `seekToBeginning()` and `seekToEnd()` is switched to `addAndGet()`, it doesn't guarantee that the positions have been reset when the method exits, does it?

Hi @kirktrue , thanks for your detailed review.
I think `seekToBeginning` and `seekToEnd` can be guarantee by follow steps:
updateFetchPositions[[1]](https://github.com/apache/kafka/blob/6744a718c2c177c7d462b231ea5e476d98f6eb38/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L252) -> updatePositionsWithOffsets[[2]](https://github.com/apache/kafka/blob/6744a718c2c177c7d462b231ea5e476d98f6eb38/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L290) -> initWithPartitionOffsetsIfNeeded[[3]](https://github.com/apache/kafka/blob/6744a718c2c177c7d462b231ea5e476d98f6eb38/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L347) -> resetPositionsIfNeeded[[4]](https://github.com/apache/kafka/blob/6744a718c2c177c7d462b231ea5e476d98f6eb38/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L482) -> sendListOffsetsRequestsAndResetPositions[[5]](https://github.com/apache/kafka/blob/6744a718c2c177c7d462b231ea5e476d98f6eb38/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L686)

So I think the `seekToEnd` and `seekTobegginneing` will wait until get responses, does it make sense?